value/matrix.go

```diff
@@ -544,3 +544,97 @@ func (m *Matrix) sel(c Context, v Vector) *Matrix {
 
 	return NewMatrix(shape, result)
 }
+
+// take returns v take m.
+func (m *Matrix) take(c Context, v Vector) *Matrix {
+	// Extend short vector to full rank using shape, like in APL.
+	if len(v) > m.Rank() {
+		Errorf("take: bad length %d for shape %s", len(v), NewIntVector(m.Shape()))
+	}
+	if len(v) < m.Rank() {
+		ext := make(Vector, m.Rank())
+		copy(ext, v)
+		for i := len(v); i < m.Rank(); i++ {
+			ext[i] = Int(m.Shape()[i])
+		}
+		v = ext
+	}
+
+	// All lhs values must be small integers in range for m's shape.
+	// Compute new shape.
+	shape := make([]int, m.Rank())
+	count := int64(1)
+	for i, x := range v {
+		y, ok := x.(Int)
+		if !ok {
+			Errorf("take: left operand must be small integers")
+		}
+		if y < 0 {
+			y = -y
+		}
+		if y > Int(m.Shape()[i]) {
+			Errorf("take: left operand %v out of range for %d in shape %v", x, m.Shape()[i], NewIntVector(m.Shape()))
+		}
+		shape[i] = int(y)
+		count *= int64(y)
+	}
+
+	result := make(Vector, 0, count)
+	result = appendTake(result, v, m.Data(), m.Shape())
+	return NewMatrix(shape, result)
```

Comment:
i think we need .shrink() here and maybe elsewhere in the new code. if take returns a scalar, you want a scalar, not a matrix here. at least, that's what the other operators do, perhaps not quite consistently. i'm trying to avoid degenerate values

Reply:
APL seems not to do the shrinking. Over on tryapl.org:

```
      ⍴1↑3 4⍴⍳12
1 4
      ⍴1 1↑3 4⍴⍳12
1 1
```

And with Ivy with these changes:

```
rho 1 take 3 4 rho iota 12
1 4

rho 1 1 take 3 4 rho iota 12
1 1
```

If shrink were applied, these would disagree. 

Without appealing to APL, I've found that it's very difficult to write programs
when you can't predict the rank of an expression. If you have

    rho n take 3 4 rho iota 12

and that's usually an (n 4) except it ends up a (4) when n == 1,
that's quite difficult to handle in code. 

Or with vectors, suppose you are splitting a list into pieces 
and trying to process two halves, like you know (rho u) >= 2
and process `v = ((rho u) div 2) take u` and `w = ((rho u) div 2) drop u`.
Today, thankfully, Ivy treats even 1-element results as vectors, not scalars.
That means that v[1] and w[1] are always valid.
If these vectors devolved into scalars, they'd stop being valid.

I ran into this with indexing too. If you have a vector v of indexes
and you use x[v], you really want to get back a vector of results,
no matter the length of v. The use of shrink means that if v
happens to have length 1, then you get a scalar. That means 
x[v][1] fails, but only in the case when v has length 1.
I definitely have had to resort to using take instead of indexing
in Ivy to avoid this problem.

Back over to tryapl.org, indexing by a vector
always returns a vector, presumably to avoid this kind of problem:

```
      x ← 1 2 3
      x[1]
1
      x[1⍴1]
1
      ⍴x[1]

      ⍴x[1⍴1]
1
```

(In APL ⍴ always returns a vector, with zero length for a scalar.)

There's a bigger conversation here about figuring out what the right
rules are for when, if ever, shrink should be applied, but take and drop
on vectors don't shrink today, which in practice works well, and I don't
think they should shrink on matrices either. That is, the rule they follow
is that they are rank-preserving operators, as are the others I added.

Reply:
Thanks for the explanation. There are cases where I've used shrink, and some of them might be right (matrix shrunk to vector for instance).

Appealing to APL is fine, but this isn't APL (and neither is dyalog, actually).

But OK.